### PR TITLE
Attach new db

### PR DESCRIPTION
**Pull request by acquamarin**
This PR allows user to introduce a remote kuzu database.
To load an attached kuzu database, users must have to install/load httpfs extension first by:
```
install httpfs;
load extension httpfs;
```
Then they can attach using:
```
attach 'https://testdb' (dbtype 'kuzu')
```
Finally, they can query any tables in testdb as querying local database.
```
match (p:person) return p
```

### CODE REVIEW

File: dataset/databases/tinysnb/.lock
Comment (ray6080) [inline dataset/databases/tinysnb/.lock:1]: hmm do we need this file in test dataset?

File: src/common/file_system/file_system.cpp
Comment (ray6080) [inline src/common/file_system/file_system.cpp:1]: Why removing `const`?

File: src/main/attached_database.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+AttachedKuzuDatabase::AttachedKuzuDatabase(std::string dbPath, std::string dbName,
+    std::string dbType, ClientContext* clientContext)
+    : AttachedDatabase{std::move(dbName), std::move(dbType), nullptr /* catalog */} {
+    auto vfs = clientContext->getVFSUnsafe();
+    auto path = vfs->expandPath(clientContext, dbPath);
+    if (!vfs->fileOrPathExists(path)) {
+        throw common::RuntimeException(common::stringFormat(
+            "Cannot attach a remote kuzu database due to invalid path: {}.", dbPath));
+    }
```
Comment (ray6080) [inline src/main/attached_database.cpp:25]: Can u also check if the wal file is empty? We should not replay wal to recover on read only database.

File: src/include/main/client_context.h [hunk 2]
```diff
@@ -171,6 +178,8 @@ class ClientContext {
     std::unique_ptr<common::RandomEngine> randomEngine;
     // Attached database.
     Database* database;
+    // Default database.
+    AttachedKuzuDatabase* defaultDatabase;
```
Comment (ray6080) [inline src/include/main/client_context.h:182]: These two names `database` and `defaultDatabase` are confusing. It's hard to tell what's the difference.

In my understanding, `defaultDatabase` should be renamed to `attachedDatabase`, and `database` should be `defaultDatabase`. Or we can also call them `localDatabase` and `remoteDatabase`, as attach can only apply to httpfs for now.
Also update the comments if you do the renaming.

File: src/main/client_context.cpp [hunk 3]
```diff
@@ -491,6 +522,21 @@ void ClientContext::commitUDFTrx(bool isAutoCommitTrx) {
     }
 }
 
+bool ClientContext::canExecuteWriteQuery() {
+    if (dbConfig.readOnly) {
+        return false;
+    }
+    // Note: we can only attach a remote kuzu database in read-only mode and only one
+    // remote kuzu database can be attached. If there is one remote
```
Comment (ray6080) [inline src/main/client_context.cpp:530]: Finish the sentence here.

File: src/main/client_context.cpp [hunk 4]
```diff
@@ -491,6 +522,21 @@ void ClientContext::commitUDFTrx(bool isAutoCommitTrx) {
...
+    auto dbManager = database->databaseManager.get();
+    for (auto& attachedDB : dbManager->getAttachedDatabases()) {
+        if (attachedDB->getDBType() == common::ATTACHED_KUZU_DB_TYPE) {
```
Comment (ray6080) [inline src/main/client_context.cpp:533]: hmm should we do the check and throw exception during attaching? so users have to specify read only for attached remote kuzu db. that's more user friendly.

**Review by ray6080 — APPROVED** [review]

File: src/common/file_system/file_system.cpp
Comment (acquamarin) [reply src/common/file_system/file_system.cpp:1]: For httpfilesystems, they are going to open the actual file to test whether the file exists or not.
So i have to remove the const keyword

File: dataset/databases/tinysnb/.lock
Comment (acquamarin) [reply dataset/databases/tinysnb/.lock:1]: removed

File: src/main/client_context.cpp (hunk 4 above)
Comment (acquamarin) [reply src/main/client_context.cpp:533]: I don't think so, we can mention this in the doc so users know that they can only execute read-only queries inside attached database.

**Comment by giusedroid** [comment]
This is beautiful <3 I'll start tinkering with it soon!

**Pull request merged**